File: class25/java/TCC2502-Otimizador.java
// TCC2502-Otimizador.java | Aula 25
// Otimizador didático: constant folding e eliminação de código morto.
// Espelho Java do TCC2502-Otimizador.py (mesmas entradas e saídas).
import java.util.*;

public class TCC2502Otimizador {

    static boolean ehNum(String s) {
        try { Double.parseDouble(s); return true; }
        catch (NumberFormatException e) { return false; }
    }

    static Double aplicarOp(double a, String op, double b) {
        switch (op) {
            case "+": return a + b;
            case "-": return a - b;
            case "*": return a * b;
            case "/": return b != 0 ? a / b : null;
            default: return null;
        }
    }

    static List<String> constantFolding(List<String> tac) {
        Map<String, Double> valores = new HashMap<>();
        List<String> otimizado = new ArrayList<>();
        for (String linha : tac) {
            String[] parts = linha.split(" = ", 2);
            if (parts.length != 2) { otimizado.add(linha); continue; }
            String dst = parts[0], expr = parts[1];
            String[] tokens = expr.split(" ");
            if (tokens.length == 3 && ehNum(tokens[0]) && ehNum(tokens[2])) {
                double a = Double.parseDouble(tokens[0]);
                double b = Double.parseDouble(tokens[2]);
                Double r = aplicarOp(a, tokens[1], b);
                if (r != null) {
                    valores.put(dst, r);
                    otimizado.add(dst + " = " + r + "  (constant folding)");
                    continue;
                }
            }
            for (int i = 0; i < tokens.length; i++) {
                if (valores.containsKey(tokens[i])) {
                    tokens[i] = String.valueOf(valores.get(tokens[i]));
                }
            }
            otimizado.add(dst + " = " + String.join(" ", tokens));
        }
        return otimizado;
    }

    static List<String> deadCodeElimination(List<String> tac) {
        Set<String> usados = new HashSet<>();
        for (String linha : tac) {
            String[] parts = linha.split(" = ", 2);
            if (parts.length == 2) {
                for (String token : parts[1].split(" ")) {
                    if (token.matches("[a-zA-Z_]\\w*") || token.startsWith("t")) {
                        usados.add(token);
                    }
                }
            }
        }
        List<String> otimizado = new ArrayList<>();
        for (String linha : tac) {
            String[] parts = linha.split(" = ", 2);
            if (parts.length == 2) {
                String dst = parts[0];
                if (dst.startsWith("t") && !usados.contains(dst)) continue;
            }
            otimizado.add(linha);
        }
        return otimizado;
    }

    public static void main(String[] args) {
        List<String> tac = Arrays.asList(
            "t1 = 3 + 4",
            "t2 = t1 * 2",
            "t3 = 10 - 5",
            "x = t2 + t3",
            "t4 = 100 * 0"
        );

        System.out.println("=== TAC Original ===");
        for (String l : tac) System.out.println("  " + l);

        System.out.println("\n=== Após Constant Folding ===");
        List<String> cf = constantFolding(tac);
        for (String l : cf) System.out.println("  " + l);

        System.out.println("\n=== Após Dead Code Elimination ===");
        List<String> dce = deadCodeElimination(cf);
        for (String l : dce) System.out.println("  " + l);
    }
}

File: class25/python/TCC2502-Otimizador.py
#!/usr/bin/env python3
"""TCC2502-Otimizador.py | Aula 25
Otimizador didático: constant folding e eliminação de código morto.
"""
from typing import List


def constant_folding(tac: List[str]) -> List[str]:
    """Constant folding: avalia subexpressões constantes no TAC."""
    valores = {}
    otimizado = []
    for linha in tac:
        parts = linha.split(" = ", 1)
        if len(parts) != 2:
            otimizado.append(linha)
            continue
        dst, expr = parts
        tokens = expr.split()
        # t1 = 3 + 4 → t1 = 7
        if len(tokens) == 3 and _eh_num(tokens[0]) and _eh_num(tokens[2]):
            a, op, b = float(tokens[0]), tokens[1], float(tokens[2])
            resultado = _aplicar_op(a, op, b)
            if resultado is not None:
                valores[dst] = resultado
                otimizado.append(f"{dst} = {resultado}  (constant folding)")
                continue
        # Substituir variáveis conhecidas
        for i, t in enumerate(tokens):
            if t in valores:
                tokens[i] = str(valores[t])
        otimizado.append(f"{dst} = {' '.join(tokens)}")
    return otimizado


def dead_code_elimination(tac: List[str]) -> List[str]:
    """Eliminação de código morto: remove atribuições cujo destino não é usado."""
    usados = set()
    for linha in tac:
        parts = linha.split(" = ", 1)
        if len(parts) == 2:
            for token in parts[1].split():
                if token.isidentifier() or token.startswith("t"):
                    usados.add(token)
    # Manter atribuições a variáveis usadas ou não-temporárias
    otimizado = []
    for linha in tac:
        parts = linha.split(" = ", 1)
        if len(parts) == 2:
            dst = parts[0]
            if dst.startswith("t") and dst not in usados:
                continue  # código morto
        otimizado.append(linha)
    return otimizado


def _eh_num(s: str) -> bool:
    try:
        float(s)
        return True
    except ValueError:
        return False


def _aplicar_op(a: float, op: str, b: float):
    ops = {"+": a + b, "-": a - b, "*": a * b, "/": a / b if b != 0 else None}
    return ops.get(op)


if __name__ == "__main__":
    tac = [
        "t1 = 3 + 4",
        "t2 = t1 * 2",
        "t3 = 10 - 5",
        "x = t2 + t3",
        "t4 = 100 * 0",  # código morto (x não usa t4)
    ]

    print("=== TAC Original ===")
    for linha in tac:
        print(f"  {linha}")

    print("\n=== Após Constant Folding ===")
    cf = constant_folding(tac)
    for linha in cf:
        print(f"  {linha}")

    print("\n=== Após Dead Code Elimination ===")
    dce = dead_code_elimination(cf)
    for linha in dce:
        print(f"  {linha}")
